core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/waiters/AsyncWaiterHandler.java

```diff
@@ -0,0 +1,118 @@
+     * Execute the provided async polling function
+     */
+    CompletableFuture<WaiterResponse<T>> execute(Supplier<CompletableFuture<T>> asyncPollingFunction) {
+        log.info(() -> "starting to execute");
+        CompletableFuture<WaiterResponse<T>> future = new CompletableFuture<>();
+        doExecute(asyncPollingFunction, future);
+        return future;
+    }
+
+    private void doExecute(Supplier<CompletableFuture<T>> asyncPollingFunction, CompletableFuture<WaiterResponse<T>> future) {
+        ++attemptNumber;
+        log.info(() -> "runOnce" + attemptNumber);
+        runAsyncPollingFunction(asyncPollingFunction, future);
+    }
+
+    private void runAsyncPollingFunction(Supplier<CompletableFuture<T>> asyncPollingFunction,
+                                         CompletableFuture<WaiterResponse<T>> future) {
+        asyncPollingFunction.get().whenComplete((response, exception) -> {
+            Optional<WaiterState> waiterState = handlerHelper.nextWaiterStateIfMatched(response, exception);
+
+            if (waiterState.isPresent()) {
+                log.info(() -> "waiter state " + waiterState.get());
+                WaiterState state = waiterState.get();
+                switch (state) {
+                    case SUCCESS:
+                        future.complete(handlerHelper.createWaiterResponse(response, exception, attemptNumber));
+                        break;
+                    case RETRY:
+                        maybeRetry(asyncPollingFunction, future);
+                        break;
+                    case FAILURE:
+                        future.completeExceptionally(SdkClientException.create("A waiter acceptor was matched and transitioned "
+                                                                               + "the waiter to failure state"));
+                        break;
+                    default:
+                        future.completeExceptionally(new UnsupportedOperationException());
+                }
+            } else {
+                if (exception != null) {
+                    future.completeExceptionally(SdkClientException.create("An exception was thrown and did not match any "
+                                                                           + "waiter acceptors"));
+                } else {
+                    // default to retry if there's no acceptor matched for the response
+                    maybeRetry(asyncPollingFunction, future);
+                }
+            }
+        });
+    }
+
+    private void maybeRetry(Supplier<CompletableFuture<T>> asyncPollingFunction,
+                            CompletableFuture<WaiterResponse<T>> future) {
+        if (attemptNumber >= pollingStrategy.maxAttempts()) {
+            future.completeExceptionally(SdkClientException.create("The waiter has exceeded max retry attempts: " +
+                                                                   pollingStrategy.maxAttempts()));
+            return;
+        }
```

Comment:
Could we make this class thread-safe and reusable by just passing through the attemptNumber on each `doExecute` call? This seems like unnecessary mutability.